### PR TITLE
Fix lazy_property compatibility with dynamo

### LINKED ISSUE #135344: dynamo fails to wrap distributions with lazy property
### 🐛 Describe the bug

To reproduce:
```python
import torch.distributions


logits = torch.randn(10)
def make_categorical(logits):
    return torch.distributions.Categorical(logits=logits).sample()
torch.compile(make_categorical, fullgraph=True)(logits)
```
Error:
```
  File "/Users/vmoens/venv/rl/lib/python3.11/site-packages/torch/_dynamo/symbolic_convert.py", line 3011, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/vmoens/venv/rl/lib/python3.11/site-packages/torch/_dynamo/symbolic_convert.py", line 3061, in inline_call_
    raise ArgsMismatchError(  # noqa: B904
torch._dynamo.exc.ArgsMismatchError: too many positional arguments.
  func = '__get__' /Users/vmoens/venv/rl/lib/python3.11/site-packages/torch/distributions/utils.py:145, args = [<class 'torch.distributions.utils.lazy_property'>, <class 'torch.distributions.utils.lazy_property'>, <class 'torch.distributions.categorical.Categorical'>, <class 'type'>], kwargs = {}
```

### Versions

nightly

cc @ezyang @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames @rec

### PR DESCRIPTION
**Pull request by vmoens**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138080

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

**Comment by Skylion007** [comment]
No tests for this?

**Comment by vmoens** [comment]
This solution isn't truly amazing, especially because it breaks how `lazy_property` really works.

The current implementations dynamically swaps `type(self).attr` with `self.attr` in a way that will be confusing for dynamo whatever we do.

The main issue is that it seems we can make it work when the distribution is built within the compiled region or outside but not both.

I would argue that `lazy_property` does something very fishy at its core. The `distributions` codebase makes multiple checks in `dist.__dict__` (fixed in this PR) which I also assume will fail when compiled - the idea is that if `probs` is in `__dict__` then it's not lazy so it can be queried. It's an indirect way of checking if the `probs` attribute has been set. Traditionally, the pythonic way of doing this would be to have a property that lazily gets `_probs` instead.

But making this change will break people's code if they rely on the behavior of `lazy_property`, and this is a public feature so we may want to be careful with it.

**Comment by vmoens** [comment]
@fritzo would you give us your thoughts on the lazy_property refactoring?

File: torch/distributions/utils.py [hunk 1]
```diff
@@ -130,25 +131,36 @@ def probs_to_logits(probs, is_binary=False):
...
-        with torch.enable_grad():
-            value = self.wrapped(instance)
-        setattr(instance, self.wrapped.__name__, value)
-        return value
+    def __init__(self, fget=None, fset=None, fdel=None, doc=None):
+        wrapped = fget
+        assert fset is None
+        assert doc is None
+        wrapped = wrapped
+        name = "_" + wrapped.__name__
```
Comment (fritzo) [inline torch/distributions/utils.py:147]: It would be safer and slightly clearer to set `name = "_lazy_" + wrapped.__name__`.  I just worry that among the thousands of user-defined distributions in the wild, many already have members with names like `._logits` and `._scale` that would conflict with a mere "_" prefix, and I'd bet there will be zero conflicts in the wild with a "_lazy_" prefix.

File: torch/distributions/utils.py [hunk 2]
```diff
@@ -130,25 +131,36 @@ def probs_to_logits(probs, is_binary=False):
...
-            return _lazy_property_and_property(self.wrapped)
-        with torch.enable_grad():
-            value = self.wrapped(instance)
-        setattr(instance, self.wrapped.__name__, value)
-        return value
+    def __init__(self, fget=None, fset=None, fdel=None, doc=None):
+        wrapped = fget
+        assert fset is None
+        assert doc is None
+        wrapped = wrapped
```
Comment (fritzo) [inline torch/distributions/utils.py:146]: vestigial?

**Review by fritzo — COMMENTED** [review]
Sorry for any difficulty the original `lazy_property` may have caused.

Generally seems reasonable.

Since this PR would break at least 6 places in Pyro, checking `if ... in self.__dict__`, and probably 10s of other places in the wild, could we think through a migration plan?  Like, is there a two-release version of this PR that first replaces the checks to be compatible with both `lazy_property` implementations, then a second step that changes `lazy_property`?  That way downstream users like Pyro could incorporate part 1 before being broken by part 2.

**Comment by vmoens** [comment]
> Since this PR would break at least 6 places in Pyro, checking if ... in self.__dict__, and probably 10s of other places in the wild, could we think through a migration plan?

Yeah I think we should not go ahead with these changes.
After all, `lazy_property` is a bit shady but it's using valid python behaviours so it should be up to dynamo to understand what is going on there.

@anijain2305 to recap:
We have an issue which is that, currently, dynamo works fine with `lazy_property` as long as the distribution is built out of the compiled region. When it's built within, we hit an issue in user_defined.py
https://github.com/pytorch/pytorch/blob/1d6932937e35e9f7f7686ce87f1f49e3e30c226e/torch/_dynamo/variables/user_defined.py#L1113-L1115
that this PR fixes
https://github.com/pytorch/pytorch/blob/121472af5872c641b00660bc14d4356fa2dc61c8/torch/_dynamo/variables/user_defined.py#L1113-L1115

Once that is done, the opposite happens: a dist built outside of the compiled region makes dynamo unhappy because `lazy_property` messes up the attributes of the class and instance. The contrast between these two issues is highlighted by the new test class in this PR.

There is an additional issue which is that, since `lazy_property` will write stuff in `instance.__dict__` that wasn't there before, checking the presence of an attribute in `__dict__` tells you if the `__get__` has been accessed or not. But dynamo doesn't really like that and these checks cause graph breaks. They're ubiquitious in `torch.distributions` and also in Pyro from what I understand. Again, it's a bit shady as a behaviour but it's valid python code.

One solution I've found without much dynamo dark magic was to fix lazy_property to make something a bit more traditional:
lazy_property will cache the value in `__dict__["_atrribute"]` instead of `__dict__["atrribute"]`, that way we keep the property name and the attribute name separated. That makes dynamo happy, and that allows us to replace the `if attribute in self.__dict__` by `hasattr(self, "_attribute")` which is arguably a bit cleaner. But that breaks the way people use `lazy_property` so we probably don't want to go down that road.

In summary, I think the actionables are:
1. We need to make the change in `user_defined.py` proposed by this PR work. To do that, the only blocker right now is to make dynamo understand what lazy_property is doing. If we fix that, we're only left with the issue that we're checking for a call to `lazy_property.__get__` indirectly by looking at `__dict__`.
2. To solve this problem, we could add a marker in `lazy_property` that tells your class that this has been set (like `setattr(instance, name + "_set", True)`. Not super clean either but probably enought to make dynamo happy and have a more direct way of checking if lazy_property did its job.
